### PR TITLE
Implement KTO

### PR DESCRIPTION
**Pull request by soumye**
# What does this PR do ?

Implement the [KTO](https://arxiv.org/abs/2402.01306) algorithm 

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [X] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [X] Does the trainer resume and restore model state all states?
- [X] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [ ] Does the trainer support `max_steps=-1` and `validation`?
- [ ] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [ X] Does the trainer have proper logging?

### CODE REVIEW

**Comment by trias702** [comment]
On a first look, am a bit confused, it looks like you modified `nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py` to handle KTO by setting `preference_loss` in the config and adding the KTO loss [here](https://github.com/NVIDIA/NeMo-Aligner/blob/ff303cb481b75a192598a9e3483f47e65fef2239/nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py#L294). If this is the case, and it works correctly via the change to `megatron_gpt_dpo_model.py`, then why do we even need `nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py`?

**Comment by soumye** [comment]
So there are two versions of DPO. The one in 
`nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py` is just a modification of the DPO loss and uses pairwise preference data with both chosen and rejected. The full version of KTO which works with binary preference data ie (promt,thumps up or down) is implemented in `nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py?`
I can remove the weaker version as strictly its functionality is all in the full version of KTO.

**Comment by ertkonuk** [comment]
> So there are two versions of KTO. The one in `nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py` is just a modification of the DPO loss and uses pairwise preference data with both chosen and rejected. The full version of KTO which works with binary preference data ie (prompt,thumps up or down) is implemented in `nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py` I can remove the weaker version as strictly its functionality is all in the full version of KTO.

I think we should remove the one in megatron_gpt_dpo_model.py. It complicates things and leads to a confusing design. As you said, we can still use the one in megatron_gpt_kto_model.py if our data is pairwise.

**Comment by soumye** [comment]
> > So there are two versions of KTO. The one in `nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py` is just a modification of the DPO loss and uses pairwise preference data with both chosen and rejected. The full version of KTO which works with binary preference data ie (prompt,thumps up or down) is implemented in `nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py` I can remove the weaker version as strictly its functionality is all in the full version of KTO.
> 
> I think we should remove the one in megatron_gpt_dpo_model.py. It complicates things and leads to a confusing design. As you said, we can still use the one in megatron_gpt_kto_model.py if our data is pairwise.

Sounds good. I can remote it then.

File: nemo_aligner/data/nlp/datasets.py [hunk 1]
```diff
@@ -353,6 +353,91 @@ def __getitem__(self, idx):
...
+        labels = ([-100] * prompt_len) + sample[prompt_len:]
+        # Separate the response from the prompt
+        response = sample[prompt_len:]
+        preference = 1 if payload["preference"] == "chosen" else 0
+
+        assert sample[0:prompt_len] == prompt, "the tokenizer for KTO has merged tokens between prompt and response"
+
+        assert (
+            sample_len <= self.seq_length
+        ), "tokenized text exceeds max seq len! truncate your data in preprocessing prior to KTO training"
```
Comment (trias702) [inline nemo_aligner/data/nlp/datasets.py:429]: I would recommend you amend this part to do the truncation automatically similar to how we now do it for DPO: https://github.com/NVIDIA/NeMo-Aligner/blob/f1fa2dcaee9b19deee0d2d33b9b6e3251ab7bac7/nemo_aligner/data/nlp/datasets.py#L335

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 2]
```diff
@@ -0,0 +1,414 @@
...
+
+    def prepare_for_validation_step(self):
+        prepare_for_validation_step(self)
+
+    def finish_validation_step(self):
+        finish_validation_step(self)
+
+    @torch.no_grad()
+    def get_logprob_batch(self, batch):
+        seq_length = batch["samples"].shape[1]
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:362]: Add: `batch_size = batch["chosen"].shape[0]`

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 3]
```diff
@@ -0,0 +1,414 @@
...
+
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:363]: Add: `num_microbatches = divide(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size)`

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        data_iter = get_iterator_k_split(batch, get_num_microbatches())
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:364]: Should be: `data_iter = get_iterator_k_split(batch, num_microbatches)`

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+        set_sync_funcs(self, forward_only=True)
+
+        fwd_bwd_function = get_forward_backward_func()
+
+        logprobs_list = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(logprobs_only=True),
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=get_num_microbatches(),
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:373]: Change to: `num_microbatches=num_microbatches`

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+            forward_only=True,
+            seq_length=seq_length,
+            micro_batch_size=self.cfg.kto.log_prob_forward_micro_batch_size,
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:376]: Change to: `micro_batch_size=self.cfg.kto.log_prob_forward_micro_batch_size * 2,`

**Review by trias702 — CHANGES_REQUESTED** [review]
LGTM, just a few small changes

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py (hunk 6 above)
Comment (soumye) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:376]: In KTO there shoulndn't be `* 2`  in the `micro_batch_size` as there is only one response per prompt over which we compute the loss.

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py (hunk 3 above)
Comment (soumye) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:363]: In KTO as there is only one response per prompt over which we compute the loss. So `self.cfg.kto.log_prob_forward_micro_batch_size=1` anyways so it wouldn't matter. But changing it for consistency with DPO code

File: nemo_aligner/data/nlp/datasets.py (hunk 1 above)
Comment (soumye) [reply nemo_aligner/data/nlp/datasets.py]: Added

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py (hunk 3 above)
Comment (trias702) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:365]: Ah yes, that's a fair point, in that case, no need to change it, but just a comment to remind people it's just a single response per prompt and not like DPO

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+    def finish_validation_step(self):
+        finish_validation_step(self)
+
+    @torch.no_grad()
+    def get_logprob_batch(self, batch):
+        seq_length = batch["samples"].shape[1]
+        batch_size = batch["samples"].shape[0]
+
+        # num_microbatches = get_num_microbatches()
+        num_microbatches = divide(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size)
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:367]: Should be: `divide(batch_size, min(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size))`

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py [hunk 8]
```diff
@@ -0,0 +1,418 @@
...
+        fwd_bwd_function = get_forward_backward_func()
+
+        logprobs_list = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(logprobs_only=True),
+            data_iterator=data_iter,
+            model=self.model,
+            num_microbatches=num_microbatches,
+            forward_only=True,
+            seq_length=seq_length,
+            micro_batch_size=self.cfg.kto.log_prob_forward_micro_batch_size,
```
Comment (trias702) [inline nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py:380]: Should be: `micro_batch_size=min(batch_size, self.cfg.kto.log_prob_forward_micro_batch_size),`

Just a guard in case someone enters a value too high for the forward mbs

**Review by trias702 — CHANGES_REQUESTED** [review]
LGTM, just one last item

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py (hunk 8 above)
Comment (soumye) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py]: Good idea.

File: nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py (hunk 7 above)
Comment (soumye) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_kto_model.py]: done

**Review by trias702 — APPROVED** [review]
LGTM, thanks
